src/sentry/ingest/consumer_v2/ingest.py

```diff
@@ -0,0 +1,64 @@
+from typing import Any, Mapping
+
+import msgpack
+from arroyo.backends.kafka.consumer import KafkaPayload
+from arroyo.types import Message
+
+from sentry.ingest.ingest_consumer import (
+    _load_event,
+    _store_event,
+    process_attachment_chunk,
+    process_individual_attachment,
+    process_userreport,
+)
+from sentry.models import Project
+from sentry.utils import metrics
+
+DecodedMessage = Mapping[str, Any]
+
+
+def process_ingest_message(message: Message[KafkaPayload]) -> None:
+    """
+    Processes a single Kafka Message containing an Event payload.
+
+    This does:
+    - Decode the Kafka payload which is in msgpack format and has a bit of
+      metadata like `type` and `project_id`.
+    - Fetch the corresponding Project from cache.
+    - Further processing by `type`:
+    ...
+    - Decode the actual event payload which is in JSON format and perform some
+      initial loadshedding on it.
+    - Store the JSON payload in the event processing store, and pass it on to
+      `preprocess_event`, which will schedule a followup task such as
+      `symbolicate_event` or `process_event`.
+    """
+
+    raw_payload = message.payload.value
+    message: DecodedMessage = msgpack.unpackb(raw_payload, use_list=False)
+
+    message_type = message["type"]
+    project_id = message["project_id"]
+
+    # TODO: we are only fetching a single project, so should we use a different metric?
+    with metrics.timer("ingest_consumer.fetch_projects"):
+        project = Project.objects.get_from_cache(id=project_id)
```

Comment:
Fetching single projects at a time adds quite a bit of latency. We should test this (as this is the most straight forward implementation). If this turns out to be too large in latency, then we can instead:
- collect messages into batches
- add a task step that runs `get_from_cache_many`
- add a multiprocessing task that does the rest

Reply:
If this is only about latency, I don’t think that matters too much, as at least events will be parallelized. Maybe even going through all the ceremony of batching things might add more latency than that.